### ISSUE TITLE
 with TLS verification

### ISSUE THREAD

**Issue by deuch** [fabiolb/fabio#362]
Hello,

We'are using Fabio 1.5.2 and we have an issue

**Comment by magiconair** [comment]
Hi @deuch, this is what I understand so far:

```
internet -> https://foo.com/bar -> fabio -> https://1.2.3.4/bar (Host: foo.com)
```

The cert is for `foo.com` but not for `1.2.3.4` and setting the `Host` header isn't sufficient. I would need to dig but can't you add 1.0.0.0/8 to the cert?

**Comment by deuch** [comment]
Hello,

Our fabio are not exposed to internet so it's more like that:

Load Balancer (Appliance) https://foo.com/bar --> Fabio (4 instances at least) --> https://1.2.3.4/bar

We can not use wildcard for our certificates (forbidden by the security) in a banking context.

Fabio and services are containers running on the same platform. For each application, we deploy a Load Balancer (Appliance), 4 Fabio instances on  a dedicated overlay networks and the services are connected to this overlay network.
On the same platform we have multiple time this setup for each application and environment (dev, int, uat ...) So i can not use a 10.0.0.0/8 wildcard certificate (because it can not be generated by our PKI) and it will break the multi tenancy of the platform.

So to be sure that a service is served by the right Fabio (or to ensure that Fabio serve the right service), it will be a good thing to check the CN of the certificate backend and not it's IP. Of course it's not the normal behaviour and must be an option for some use case like mine. The normal behaviour is to check the CN/SAN of the back-end with the back-end FQDN registered in consul (for my use case an IP).

With container, generate a certificate each time a container is created its a too heavy operation and difficult to maintain (certificate revocation will be a nightmare ...)

**Comment by magiconair** [comment]
Would the `host=dst` from #294 help? (I really need to work on the docs)

**Comment by deuch** [comment]
I'm not sure to understand the behaviour of this : With host=dst, what will be the header when connecting the upstream ? The upstream hostname/ip ? Or the hostname header coming to Fabio ?

**Comment by magiconair** [comment]
`host=dst` is for the reverse proxy case. Let me think about this a bit more.

**Comment by magiconair** [comment]
`host=dst` will effectively swap the hostname in the upstream request:

     https://a.com/foo -> fabio -> https://b.com/foo 'Host: b.com' (with host=dst)

`host=dst` will use the consul service address for the `Host` header and the target URL. If your container has a cert for its ip and that cert is trusted by fabio then this should work. But creating a cert per container is something you can't do right now.

For HTTPS, it isn't sufficient to set the `Host` header since it is transmitted after the TLS handshake. The right server name needs to be in the request URL, i.e. `https://a.com/foo` (TLS server name: `a.com`) and not `https://1.2.3.4/foo Host: a.com` (TLS server name: `1.2.3.4`).

In essence, you want fabio to make the upstream request with the original hostname (e.g. `a.com`) since the cert in the docker container contains that name. 

fabio could either spoof the DNS lookup for that request or try to establish the TCP connection first and then run the TLS handshake with the original server name (which circumvents the DNS lookup).

This would then allow you to re-use the same cert on all upstream servers. 

Does that make sense?

**Comment by deuch** [comment]
Yes it makes sense. In my use case, the TLS server name has to be the original request indeed.

I do not know what is the best thing to do. DNS spoofing or TCP connection first and TLS handshake after. What is the most secure ?

**Comment by deuch** [comment]
Hello, did you have time to try some stuff for this use case ?

**Comment by magiconair** [comment]
not yet. sorry.
